experimental/dds/tree2/src/codec/parameterized-dds.md

```diff
@@ -0,0 +1,252 @@
+
+```typescript
+const myDataObject = new DataObjectFactory("foo", MyDataObject, [
+	SharedSomething.getFactory({ formatVersion: 5 }),
+]);
+```
+
+If they later write `SharedSomething.create(runtime, { formatVersion: 4 })`, the create flow outlined above will instead use the
+factory created at data object registry construction time.
+
+There is another, more subtle issue with this strategy: how should one implement a `SharedSomethingFactory` which takes in some parameters?
+
+```typescript
+class SharedSomethingFactory {
+	public constructor(private readonly options: SharedSomethingOptions) {}
+
+	public load(
+		runtime: IFluidDataStoreRuntime,
+		id: string,
+		services: IChannelServices,
+		channelAttributes: Readonly<IChannelAttributes>,
+	): Promise<IChannel> {
+		const sharedSomething = new SharedSomething(
+			runtime,
+			id,
+			this.options /* uh oh: channelAttributes and this.options might disagree on snapshot format! */,
+		);
+		await sharedSomething.load(services);
+		return sharedSomething;
+	}
+
+	public create(runtime: IFluidDataStoreRuntime, id: string): IChannel {
+		const sharedSomething = new SharedSomething(document, id, this.options);
+		sharedSomething.initializeLocal();
+		return sharedSomething;
+	}
+}
+```
+
+If the channel's attributes clearly reflect its parameters (e.g. the snapshot format being used),
+then the resulting factory sometimes produces DDSes which don't match the requested format!
+
+There are essentially two ways to resolve this issue:
+
+1. Make `SharedSomethingFactory` have a parameterless constructor, and pass parameters all the way through the `create` flow (i.e. through `createChannel`, factory lookup, and factory's `create` method).
+2. Keep parameters in `SharedSomethingFactory`'s constructor, but change their semantics such that they only necessarily apply to newly created documents.
+
+Regardless of which option we pick, enough information to reify the DDS's parameters needs to be persisted in the attributes blob.
+
+In order to support data objects which can create multiple types of the same shared object with different parameters, option 2 would need to be accompanied by changes to make DDS registration take in named entries in `@fluidframework/aqueduct`, rather than assume the "key" for a DDS registry entry is that DDS's factory's `type` field.
+Option 2 would also necessarily force us away from the static `create` method commonly found on DDSes.
+
+```typescript
+// Container author code, from the perspective of a container author
+// Option 1:
+const myDataObject = new DataObjectFactory("foo", MyDataObject, [SharedSomething.getFactory()]);
+
+// later, they make a SharedSomething:
+const something = SharedSomething.create({ formatVersion: 5 });
+
+// Option 2:
+const myDataObject = new DataObjectFactory("foo", MyDataObject, [
+	["v4Something", SharedSomething.getFactory({ formatVersion: 4 })],
+	["v5Something", SharedSomething.getFactory({ formatVersion: 5 })],
```

Comment:
This example is kind of assuming there's a single configuration per factory, e.g. otherwise my factory list might have members like "v4SnapshotAttributionEnabledWithSummarizeHistorySomething"

Reply:
yeah, fair point. I don't think people should actually name their keys like this. Probably going with "foo" and "bar" is better :P.

Reply:
Well, naming aside - my point is more that it would be good to explore whether this approach scales well to the number of configs we think folks may want.

Reply:
gotcha. let's continue that discussion on https://github.com/microsoft/FluidFramework/pull/15999#discussion_r1230289988 then (seems to be that point)